haproxy-healthcheck/check.py

```diff
@@ -1,80 +1,87 @@
+password: str = os.environ["COUCHDB_PASSWORD"]
 
-def get_membership_endpoint(couchdb_url):
-    return f'http://{username}:{password}@{couchdb_url}:5984/_membership'
 
-def close_connection(connection):
-    print('Closing connection ...')
-    connection.close()
+def get_membership_endpoint(couchdb_url: str) -> str:
+    return f"http://{username}:{password}@{couchdb_url}:5984/_membership"
 
-def send_down_response(conn):
-    print('Sending response: Down (CouchDB cluster not okay) ...')
-    conn.send(b'down\n')
-    close_connection(conn)
 
-def send_up_response(conn):
-    print('Sending response: Up (Couchdb cluster okay) ...')
-    conn.send(b'up\n')
-    close_connection(conn)
+def _send_down_response(writer: asyncio.StreamWriter) -> None:
+    _send_and_close(writer, b"down\n")
+
+
+def _send_up_response(writer: asyncio.StreamWriter) -> None:
+    _send_and_close(writer, b"up\n")
 
-def threaded(conn):
-    try:
-        result = []
-        couchdb_server = couchdb_servers[0]
-        try:
-            r = requests.get(get_membership_endpoint(couchdb_server), timeout=1)
-        except Exception as e:
-            print("Exception: ", e)
-        data = r.json()
 
-        all_nodes = sorted(data['all_nodes'])
-        cluster_nodes = sorted(data['cluster_nodes'])
+def _send_and_close(writer: asyncio.StreamWriter, message: bytes) -> None:
+    print(f"Response: {message!r}")
+    writer.write(message)
+    print("Closing connection")
+    writer.close()
+
+
+async def is_healthy() -> bool:
+    """
+    Checks that all couchdb servers are part of the cluster
+    """
+    try:
+        first_couchdb_server: str = couchdb_servers[0]
+        async with httpx.AsyncClient() as client:
+            r = await client.get(
+                get_membership_endpoint(first_couchdb_server), timeout=1
+            )
+        data: dict = r.json()
+        all_nodes: List[str] = sorted(data["all_nodes"])
+        cluster_nodes: List[str] = sorted(data["cluster_nodes"])
 
         if len(all_nodes) != len(couchdb_servers):
-            print('Nodes starting up')
-            print(f'Details: all_nodes: {all_nodes}. couchdb_servers: {couchdb_servers}')
-            result.append(b'down\n')
+            print("Nodes starting up")
+            print(
+                f"Details: all_nodes: {all_nodes}. couchdb_servers: {couchdb_servers}"
+            )
+            return False
         elif all_nodes != cluster_nodes:
-            print('_membership shows not all nodes are part of Cluster')
-            print(f'Details: all_nodes: {all_nodes}. cluster_nodes: {cluster_nodes}')
-            result.append(b'down\n')
+            print("_membership shows not all nodes are part of Cluster")
+            print(f"Details: all_nodes: {all_nodes}. cluster_nodes: {cluster_nodes}")
+            return False
         else:
-            print('Everything is fine')
-            result.append(b'up\n')
-        print(result)
+            return True
+
```

Comment:
I'm thinking about refining this check to check if the cluster is running all expected servers. 
CouchDb will incorrectly list extra nodes that it finds referenced in shard maps: 

For example, I started a cluster on top of a single node data set and this happens. 
```
Details: all_nodes: ['couchdb@couchdb-1.local', 'couchdb@couchdb-2.local', 'couchdb@couchdb-3.local']. cluster_nodes: ['couchdb@127.0.0.1', 'couchdb@couchdb-1.local', 'couchdb@couchdb-2.local', 'couchdb@couchdb-3.local']
```

While we don't expect this to happen in production, when we migrate (from 3.x to clustered) or change the cluster and move shards around, we _can_ be in this situation and this prevents CouchDb from being reachable through haproxy. 

What do you think? Can we check both all_nodes and cluster_nodes include couchdb_servers? or have all_nodes be equal and cluster_nodes be inclusive?